feat(SLA): Apply SLA to any document

hrwX opened the pull request:
This PR enables setting Service Level Agreement for any doctype.

### Document Type field added to SLA doctype

![document_type](https://user-images.githubusercontent.com/24353136/120074777-17c36c80-c0bc-11eb-94f8-1c728d1a5b3a.png)

### SLA fields

On creating the SLA, an SLA section will get added to the selected doctype with the following fields (custom fields for standard doctype and docfields for custom doctype).
  - Service Level Agreement
  - Priority
  - Response By
  - Response By Variance
  - Resolution By
  - Resolution By Variance
  - First Responded On
  - On Hold Since
  - Total Hold Time
  - Agreement Status
  - SLA creation
  - Resolution Date

![custom_dt_with_sla](https://user-images.githubusercontent.com/24353136/120075479-09c31b00-c0bf-11eb-8fae-46d3feae5234.png)

### Hold Time

Hold Time is captured if the **Pause SLA On**  configuration is setup in SLA

https://user-images.githubusercontent.com/24353136/120075658-d9c84780-c0bf-11eb-9fa5-f17a9fb0dc8c.mp4

### SLA Fulfilled On

Allows you to configure statuses for SLA fulfillment

![image](https://user-images.githubusercontent.com/24353136/120075753-2c096880-c0c0-11eb-86b9-55000eb9df0e.png)

### Apply SLA for Resolution Time

For a few instances like applying SLA on leads, there is no resolution SLA required. The only metric that matters is the First Response Time. Hence, the SLA application for resolution time is now optional. If it is disabled, the SLA status is validated as Fulfilled / Failed on the basis of first response time only.

![image](https://user-images.githubusercontent.com/24353136/120075842-7db1f300-c0c0-11eb-8146-9a7f001945f9.png)

https://user-images.githubusercontent.com/24353136/120075975-22343500-c0c1-11eb-9dd8-4e76478e540c.mp4

### Hooks
   1. Applying SLA
   2. Setting hold time on communication update
   3. Setting SLA variance

Documentation PR: https://github.com/frappe/erpnext_documentation/pull/340

hrwX committed:
fix: SLA for multiple doctypes

hrwX committed:
fix: set hold time

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into multiple_slas

hrwX committed:
Merge branch 'develop' into multiple_slas

hrwX committed:
fix: remove unwanted fields

hrwX committed:
fix: add custom fields

hrwX committed:
Merge branch 'multiple_slas' of https://github.com/hrwx/erpnext into multiple_slas

hrwX committed:
fix: remove unused imports

hrwX committed:
Merge branch 'develop' into multiple_slas

hrwX committed:
fix: ignore permissions

hrwX committed:
Merge branch 'multiple_slas' of https://github.com/hrwx/erpnext into multiple_slas

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into multiple_slas

hrwX committed:
fix: sla not being applied

hrwX committed:
fix: consider user tz for sla

hrwX committed:
fix: pass user as arg

hrwX committed:
fix: convert time str to timedelta

hrwX committed:
fix: add first responded on field

hrwX committed:
fix: get timezone

hrwX committed:
fix: helper function to get user time

hrwX committed:
fix: set service_level_agreement_creation

hrwX committed:
Merge branch 'develop' into multiple_slas

deepeshgarg007 committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into multiple_slas

deepeshgarg007 committed:
fix: Codacy Issues

deepeshgarg007 committed:
fix: Codacy issues and travis

deepeshgarg007 committed:
fix: Add resolved status field in Pause SLA

deepeshgarg007 committed:
fix: Put SLA on hold for resolved status

hrwX committed:
Merge branch 'develop' into multiple_slas

hrwX committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
fix: codacy issues and patch

ruchamahabal committed:
fix: codacy issue

ruchamahabal committed:
fix: validate if doctype for SLA configuration has status field

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
fix: remove duplicate code, fix sider and translation issues

ruchamahabal committed:
fix: SLA status not getting updated

ruchamahabal committed:
fix: sider

ruchamahabal committed:
fix: Response and Resolution variance duration fieldtype

ruchamahabal committed:
fix: SLA not getting applied for custom doctypes

ruchamahabal committed:
fix(hooks): update sla variance

ruchamahabal committed:
feat: show SLA indicators

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
fix: set SLA fulfilled on Resolved

ruchamahabal committed:
fix: travis

ruchamahabal committed:
feat: Hold time for SLA on custom doctypes

ruchamahabal commented:
@hrwx Why have we removed the Employee Group field from SLA doctype in this PR? Any specific reason?

ruchamahabal committed:
fix(patch): rename enable to enabled in SLA

ruchamahabal committed:
fix: incorrect response and resolution variance calculation

ruchamahabal committed:
fix: travis and hook for updating hold time on communication

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
fix: sider issues

ruchamahabal committed:
fix: add code lost in merge conflict resolution

ruchamahabal committed:
feat: Ability to configure SLA Fulfilled On statuses

ruchamahabal committed:
fix: tests

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
test: added more tests for SLA

- custom field creation for SLA on standard dt

- docfield creation for SLA on custom dt

- SLA application

- hold time calculation

ruchamahabal committed:
fix: sider

ruchamahabal committed:
refactor: shorter functions for sla application

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
feat: Apply SLA for Resolution configuration

ruchamahabal committed:
fix: sider

ruchamahabal committed:
test: apply SLA for resolution disabled

ruchamahabal committed:
fix: show resolution time in gridview only if 'apply sla on resolution' is enabled

ruchamahabal committed:
fix: update custom dt after adding SLA fields

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

nextchamp-saqib committed:
Merge branch 'develop' into multiple_slas

nextchamp-saqib committed:
Merge branch 'develop' into multiple_slas

nextchamp-saqib commented on erpnext/support/doctype/service_level_agreement/service_level_agreement.py:
```diff
@@ -101,78 +105,169 @@ def get_service_level_agreement_priority(self, priority):
+		else:
+			self.create_custom_fields(meta, service_level_agreement_fields)
+
+	def on_trash(self):
```
Can the custom fields be deleted on trash?

ruchamahabal replied:
Unsure. IMO, deleting fields is a bit risky. If the users have some custom app or have created custom fields by themselves, they wouldn't want them to be deleted.

nextchamp-saqib committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
fix: update resolution while updating hold time only if 'apply SLA for resolution' is enabled

ruchamahabal committed:
chore: breakdown handle_hold_time function into smaller functions

ruchamahabal committed:
fix(ux): resolution time field not visible on enabling resolution application

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

ruchamahabal committed:
Merge branch 'develop' into multiple_slas

barredterra commented:
@hrwX do you remember why you decided to convert timestamps to the user timezone in the backend for this feature? It seems like all other code works with the system timezone in the backend and lets the UI/frontend convert to/from the user timezone.

We're currently [contemplating to remove this](https://github.com/frappe/frappe/pull/19933#issuecomment-1455131885).
